### PR TITLE
[SPARK-15463][SQL] support creating dataframe out of Dataset[String] for csv data

### PR DESCRIPTION
**Pull request by xwu0226**
## What changes were proposed in this pull request?

Currently only `DataFrameReader.csv(...): DataFrame` does not support converting Dataset[String] to a DataFrame. 

This PR is to add the API `DataFrameReader.csv(rdd: Dataset[String]): DataFrame`. Also in order to easily invoke the helper functions that are already implemented for csv parsing, I moved some of the private methods from `csv.DefaultSource` to `CSVRelation`.
## How was this patch tested?

A test case is added to load csv files to Dataset[String] and covert to DataFrame and check the results. 
Regression test is run.

### CODE REVIEW

**Comment by xwu0226** [comment]
@HyukjinKwon @falaki  Could you review the PR? Thanks!

**Comment by pjfanning** [comment]
Takeshi Yamamuro suggested on https://issues.apache.org/jira/browse/SPARK-15463 that the new API should take a Dataset[String] as input instead of an RDD[String]

**Comment by xwu0226** [comment]
Yes. I am adding the Dataset[String] API also. will push soon.

**Comment by HyukjinKwon** [comment]
cc @rxin 

(@xwu0226 For me, I think it might be better if the structure of CSV and JSON data sources have the same structure so that we can fix up similar issues together easily if you are going to change the structure.)

**Comment by maropu** [comment]
Do we still need the interface for RDD[String]?

**Comment by xwu0226** [comment]
@maropu The API that converts Dataset[String] to DataFrame is using the one for RDD[String]. So i am thinking it could be beneficial to provide both there?

**Comment by xwu0226** [comment]
@HyukjinKwon Let me try to understand your question.  Right now, we have `csv.DefaultSource` implementing `DataSourceRegister` and `csv.CSVRelation` contains some parsing helpers, while `json.JSONRelation.DefaultSource` implements `DataSourceRegister` and also has some parsing helpers.  Would it be better to have `csv.CSVRelation` and `csv.DefaultSource` combine, just like `json.JSONRelation`?  Do you mean something like this? Thanks!

**Comment by HyukjinKwon** [comment]
@xwu0226 Ah, for example, it seems a new method, `CSVInferSchema.inferSchemaFromRDD(..)`, seems introduced here while `CSVInferSchema.infer(...)` is already implemented. For JSON `InferSchema.infer(...)` seems responsible for both cases.
BTW, more importantly, it seems arguable to add new APIs.

**Comment by rxin** [comment]
Let's not add so many new APIs. You can just add the Dataset[String] one, since RDD[String] can be easily converted into Dataset.

**Comment by xwu0226** [comment]
@rxin OK. Thanks! will do.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala [hunk 1]
```diff
@@ -42,16 +42,23 @@ private[csv] object CSVInferSchema {
       tokenRdd: RDD[Array[String]],
       header: Array[String],
       options: CSVOptions): StructType = {
-    val startType: Array[DataType] = Array.fill[DataType](header.length)(NullType)
-    val rootTypes: Array[DataType] =
-      tokenRdd.aggregate(startType)(inferRowType(options), mergeRowTypes)
+    val structFields = if (options.inferSchemaFlag) {
```
Comment (xwu0226) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala]: This method is used in both `csv.DefaultSource` and `DataFrameReader.csv(ds: Dataset[String])`. So I refactored it here to take care both the default schema type and `inferSchemaFlag=true` cases.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala [hunk 2]
```diff
@@ -142,6 +145,75 @@ object CSVRelation extends Logging {
...
+      field.dataType match {
+        case _: ArrayType | _: MapType | _: StructType =>
+          throw new UnsupportedOperationException(
+            s"CSV data source does not support ${field.dataType.simpleString} data type.")
+        case _ =>
+      }
+    }
+  }
+
+  def getHeader(rdd: RDD[String], csvOptions: CSVOptions): Array[String] = {
```
Comment (xwu0226) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVRelation.scala]: This is also used in a few places to get the header from csv records.

**Comment by xwu0226** [comment]
@rxin Please help double check! Many thanks!!

**Comment by xwu0226** [comment]
test this please

**Comment by pjfanning** [comment]
@xwu0226 the unit tests you added seem sufficient

**Comment by xwu0226** [comment]
@yhuai @rxin Thanks!

**Comment by pjfanning** [comment]
@HyukjinKwon @rxin @falaki Would it be feasible to get this merged for Spark 2.0 release?

**Comment by rxin** [comment]
@pjfanning we are now focusing on bug fixes and stability fixes rather than adding new features.

**Comment by xwu0226** [comment]
@rxin Do you think we can revisit this feature and have it in 2.1? Thanks!

**Comment by pjfanning** [comment]
Would it be feasible to get this merged for Spark 2.1.0?

**Comment by maropu** [comment]
What's the status of this pr? It seems to be more natural that we implement `from_csv` in a similar way of `from_json` in https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/functions.scala#L2900

**Comment by maropu** [comment]
I checked the feasibility to implement `from_json` in `sql.functions`; If we move csv parser code (`CSVReader`, `CSVOptions`, ...) from `o.a.s.sql.execution.datasources.csv` to `o.a.s.sql.catalyst.csv`, we can easily implement `from_json` with a few modifications.

```
scala> val ds1 = Seq("""1,abc""").toDS()
ds1: org.apache.spark.sql.Dataset[String] = [value: string]

scala> val schema = new StructType().add("a", IntegerType).add("b", StringType)
schema: org.apache.spark.sql.types.StructType = StructType(StructField(a,IntegerType,true), StructField(b,StringType,true))

scala> val ds2 = ds1.select(from_csv($"value", schema))
ds2: org.apache.spark.sql.DataFrame = [csvtostruct(value): struct<a: int, b: string>]

scala> ds2.printSchema
root
 |-- csvtostruct(value): struct (nullable = true)
 |    |-- a: integer (nullable = true)
 |    |-- b: string (nullable = true)


scala> ds2.show
+------------------+
|csvtostruct(value)|
+------------------+
|           [1,abc]|
+------------------+
```

**Comment by xwu0226** [comment]
@maropu Thanks for the comments! It seems like adding such new datasource API in DataFrameReader is not in the priority now. That is why it has been in relatively idle state now. What you are proposing can solve similar problem, it is just that the user can not use datasource API to do it.

**Comment by maropu** [comment]
This pr seems stale and inactive. I know this kind of API changes has lower priorities now. So, how about closing this pr for now and setting `LATER` in the corresponding JIRA? Thought? cc: @rxin @xwu0226 @HyukjinKwon

**Comment by HyukjinKwon** [comment]
Actually, this feature might not be urgent as said above but IMO I like this feature to be honest. I guess the reason it was hold is that IMHO it does not look a clean fix. 

I recently refactored this code path and I have one left PR, https://github.com/apache/spark/pull/16680. After hopefully merging, there can be a easy clean fix consistently with json one within 10ish line additions, for example, something like one below in `Dataset`...

```scala
def csv(csv: Dataset[String]): DataFrame = {
  val parsedOptions: CSVOptions = new CSVOptions(extraOptions.toMap)
  val caseSensitive = sparkSession.sessionState.conf.caseSensitive
  val schema = userSpecifiedSchema.getOrElse {
    InferSchema.infer(csv, caseSensitive, parsedOptions)
  }

  val parsed = csv.mapPartitions { iter =>
    val parser = new UnivocityParser(schema, caseSensitive, parsedOptions)
    iter.flatMap(parser.parse)
  }

  Dataset.ofRows(
    sparkSession,
    LogicalRDD(schema.toAttributes, parsed)(sparkSession))
}
```

I remember there have been a quite bit of questions about this feature in spark-csv as thirdparty (and also spark-xml too).

**Comment by xwu0226** [comment]
@HyukjinKwon Thanks! After your #16680 is merged, submit a PR with the code you show above. then.

**Comment by maropu** [comment]
Yea, I also think `json` and `csv` stuffs should be consistent and they'd be better to have the same code structure and behaviour as @HyukjinKwon said. Since we do not have `DataFrameReader.json(Dataset[String])` in the current master, `functions.from_json` is the only interface for that conversion. If we add `csv(Dataset[String])` in `DataFrameReader`, do we need to also add `json(Dataset[String])` there even though we already have `functions.from_json`?

**Comment by HyukjinKwon** [comment]
Oh, I remember the answer from my previous similar question, which was that we should not add some APIs just for consistency.

I have some references about the requests for this feature but don't have ones for the others. So, I am less sure.

**Comment by maropu** [comment]
Aha, I see. Anyway, we need to keep discussion not here but the JIRA! (because this is the closed..)

**Pull request closed without merging**